File: juno_rs/strategies/macdrsi.rs
use crate::{
    strategies::{combine, Macd, MidTrend, Persistence, Rsi, Strategy},
    Advice, Candle,
};

pub struct MacdRsi {
    macd: Macd,
    rsi: Rsi,
    mid_trend: MidTrend,
    persistence: Persistence,
}

impl MacdRsi {
    pub fn new(
        macd_short_period: u32,
        macd_long_period: u32,
        macd_signal_period: u32,
        rsi_period: u32,
        rsi_up_threshold: f64,
        rsi_down_threshol: f64,
        persistence: u32,
    ) -> Self {
        Self {
            macd: Macd::new(macd_short_period, macd_long_period, macd_signal_period, 0),
            rsi: Rsi::new(rsi_period, rsi_up_threshold, rsi_down_threshol, 0),
            mid_trend: MidTrend::new(MidTrend::POLICY_IGNORE),
            persistence: Persistence::new(persistence, false),
        }
    }
}

impl Strategy for MacdRsi {
    fn update(&mut self, candle: &Candle) -> Advice {
        let macd_advice = self.macd.update(candle);
        let rsi_advice = self.rsi.update(candle);
        let advice = combine(macd_advice, rsi_advice);
        self.persistence.update(advice)
    }
}

File: juno/strategies/macdrsi.py
import operator
from decimal import Decimal

from juno import Advice, Candle
from juno.constraints import Int, Pair, Uniform

from .macd import Macd
from .rsi import Rsi
from .strategy import Meta, StrategyBase


class MacdRsi(StrategyBase):
    @staticmethod
    def meta() -> Meta:
        return Meta(
            constraints={
                ('macd_short_period', 'macd_long_period'):
                    Pair(Int(1, 100), operator.lt, Int(2, 101)),
                'macd_signal_period': Int(1, 101),
                'rsi_period': Int(1, 101),
                'rsi_up_threshold': Uniform(Decimal('50.0'), Decimal('100.0')),
                'rsi_down_threshold': Uniform(Decimal('0.0'), Decimal('50.0')),
                'persistence': Int(0, 10),
            }
        )

    _macd: Macd
    _rsi: Rsi

    def __init__(
        self,
        macd_short_period: int = 12,
        macd_long_period: int = 26,
        macd_signal_period: int = 9,
        rsi_period: int = 14,
        rsi_up_threshold: Decimal = Decimal('70.0'),
        rsi_down_threshold: Decimal = Decimal('30.0'),
        persistence: int = 0,
    ) -> None:
        self._macd = Macd(macd_short_period, macd_long_period, macd_signal_period, 0)
        self._rsi = Rsi(rsi_period, rsi_up_threshold, rsi_down_threshold, 0)
        super().__init__(
            maturity=self._macd.maturity,
            persistence=persistence
        )
        self.validate(
            macd_short_period, macd_long_period, macd_signal_period, rsi_period, rsi_up_threshold,
            rsi_down_threshold, persistence
        )

    def tick(self, candle: Candle) -> Advice:
        return Advice.combine(self._macd.update(candle), self._rsi.update(candle))
